Enable color logging on Windows via colorama

mivade opened the pull request:
This is a slight rework of #1487 which was closed when I reorganized my clone. I'm not sure what the proper way to test this would be, but I used the following script to visually check that the output does as expected:

``` python
import logging
from tornado.log import enable_pretty_logging

logger = logging.getLogger()
enable_pretty_logging()
logger.setLevel(logging.DEBUG)

logger.info("INFO")
logger.warning("WARNING")
logger.debug("DEBUG")
logger.error("ERROR")
print("normal")
```

bdarnell commented on tornado/log.py:
```diff
@@ -126,6 +135,10 @@ def __init__(self, color=True, fmt=DEFAULT_FORMAT,
             for levelno, code in colors.items():
                 self._colors[levelno] = unicode_type(curses.tparm(fg_color, code), "ascii")
             self._normal = unicode_type(curses.tigetstr("sgr0"), "ascii")
+        elif colorama is not None:
```
This also needs to be conditioned on the `color` variable and `_stderr_supports_color()`. Make this 

```
    if color and _stderr_supports_color():
        if curses is not None:
           ...
        elif colorama is not None:
           ..
        else:
            raise RuntimeError("no supported color terminal library")
```

bdarnell commented on tornado/log.py:
```diff
@@ -57,6 +64,8 @@ def _stderr_supports_color():
                 color = True
         except Exception:
             pass
+    elif colorama and os.isatty(sys.stderr.fileno()):
```
Why does this use os.isatty while the preceding check uses sys.stderr.isatty?

mivade replied:
Originally this was due to a misunderstanding on my part of #920. Using `os.isatty` means we don't need to additionally check if `sys.stderr` has the `isatty` attribute. From [`io.IOBase`](https://docs.python.org/3/library/io.html#io.IOBase), `sys.stderr` should always have a `fileno` attribute, however.

I didn't want to change the line that was already working and have something break, but it would probably make sense to just check once if `os.isatty(...)` or `sys.stderr.isatty()`. I'll rework that.

mivade commented:
Any other comments, @bdarnell?

Just pinging re: your recent post on the mailing list.

bdarnell commented on tornado/log.py:
```diff
@@ -32,12 +32,19 @@
 
 try:
+    import colorama
+    colorama.init()
```
I'm not really comfortable calling `colorama.init` at import time. It looks like it may not be idempotent, and replacing `sys.std{out,err}` is intrusive enough that there should be a more definite opt-in than "the colorama package is installed" (colorama's issue tracker shows that its approach has quite a few quirks). 

What if we leave the call to `colorama.init()` to the application and replace the `if colorama is not None` calls below with `if sys.stderr is getattr(colorama, 'wrapped_stderr', object())`? That way we'd be able to detect the use of `colorama` as an alternative indicator that ANSI color is supported (instead of `curses`'s inspection of `$TERM` and the `terminfo` database)

It looks like windows 10 also has a native ANSI-aware terminal (see tartley/colorama#104), so another approach we could take would be to ignore colorama and figure out how to detect that terminal in the absence of `curses` and `terminfo`.

mivade commented:
I agree with your objection to calling `colorama.init` explicitly. I've removed that and added language to the `LogFormatter` docstring pointing out that colorama must be initialized by the user.

I'm reluctant to rely on Windows 10 ANSI-aware terminal support for now since there is still a lot of use of Windows 7.

bdarnell commented:
Looking good, just one more thing (I think) to fix the build. Sphinx is complaining that it can't find `colorama.init` (single backticks are links in RST format, not just monospace like in markdown). Those need to be double backticks instead.

mivade commented:
Thanks for the input. I was using the `:func:` directive there, but with colorama not installed, that resulted in the error. I've replaced with double backticks instead.

bdarnell commented:
Thanks!
